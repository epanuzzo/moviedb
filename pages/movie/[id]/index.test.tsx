import React from "react";
import renderer from "react-test-renderer";
import MoviePage from "./index";
import { useDispatch, useSelector } from "react-redux";
import { useRouter } from "next/router";

jest.mock("react-redux", () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

jest.mock("../../../redux/slices/movieSlice", () => ({
  getMovieDetails: jest.fn(),
}));

jest.mock("next/router", () => ({
  useRouter: jest.fn(),
}));

describe("MoviePage", () => {
  let mockDispatch: jest.Mock;
  let mockRouter: unknown;

  beforeEach(() => {
    mockDispatch = jest.fn();
    mockRouter = {
      pathname: "/current-page",
      push: jest.fn(),
      query: { id: "1" },
    };

    (useRouter as unknown as jest.Mock).mockReturnValue(mockRouter);
    (useDispatch as unknown as jest.Mock).mockReturnValue(mockDispatch);
    (useSelector as unknown as jest.Mock).mockReturnValue({
      movies: [
        { id: 1, title: "Movie 1" },
        { id: 2, title: "Movie 2" },
      ],
      moviesDetails: {
        "1": { id: 1, title: "Movie 1", production_companies: [], genres: [] },
      },
      loading: false,
      pages: 2,
      currentPage: 1,
    });
  });

  it("renders the component", () => {
    const tree = renderer.create(<MoviePage />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("displays error message", () => {
    (useSelector as unknown as jest.Mock).mockReturnValueOnce({
      movies: [],
      moviesDetails: {},
      loading: false,
      pages: 0,
      currentPage: 0,
    });

    const tree = renderer.create(<MoviePage />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("displays loading state", () => {
    (useSelector as unknown as jest.Mock).mockReturnValueOnce({
      movies: [{ id: 1, title: "Movie 1" }],
      moviesDetails: {},
      loading: true,
      pages: 1,
      currentPage: 1,
    });

    const tree = renderer.create(<MoviePage />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
