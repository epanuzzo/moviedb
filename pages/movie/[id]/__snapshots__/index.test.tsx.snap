// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviePage displays error message 1`] = `
<div
  className="flex items-center justify-center h-90 w-full"
>
  <div
    className="text-center"
  >
    <h1
      className="text-2xl font-bold mb-4"
    >
      Movie not found
    </h1>
    <p
      className="text-gray-600"
    >
      Please check the movie ID and try again.
    </p>
    <p
      className="text-gray-600"
    >
      Use only navigation from the home page.
    </p>
    <p
      className="text-gray-600"
    >
      Reload will not work.
    </p>
  </div>
</div>
`;

exports[`MoviePage displays loading state 1`] = `
<div
  className="w-full h-full"
>
  <div
    className="flex justify-center items-center pt-4"
  >
    <div
      className="animate-spin rounded-full h-16 w-16 border-t-4 border-blue-500 border-solid"
    />
  </div>
</div>
`;

exports[`MoviePage renders the component 1`] = `
<div
  className="w-full h-full"
>
  <div
    className="bg-gray-900 text-white p-6 rounded-lg shadow-lg max-w-4xl mx-auto"
  >
    <div
      className="mb-6"
    >
      <a
        className="inline-block bg-gray-700 hover:bg-gray-600 text-white font-semibold py-2 px-4 rounded-md"
        href="/"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        ‚Üê Back to Movies list
      </a>
    </div>
    <div
      className="relative w-full h-80 bg-cover bg-center rounded-lg overflow-hidden"
      style={
        {
          "backgroundImage": "url(https://image.tmdb.org/t/p/originalundefined)",
        }
      }
    >
      <div
        className="absolute inset-0 bg-gradient-to-t from-black to-transparent opacity-90"
      />
      <div
        className="absolute top-4 left-4 bg-white text-black rounded-full flex items-center justify-center w-12 h-12"
      >
        <p
          className="text-sm font-bold"
        />
      </div>
    </div>
    <div
      className="mt-6"
    >
      <h1
        className="text-4xl font-bold"
      >
        Movie 1
      </h1>
      <p
        className="text-gray-400 italic"
      />
      <p
        className="mt-4 text-lg"
      />
    </div>
    <div
      className="mt-4 space-y-2"
    >
      0
      0
    </div>
  </div>
</div>
`;
